sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/DecryptorV2.java

```diff
@@ -0,0 +1,140 @@
+
+import com.azure.core.cryptography.AsyncKeyEncryptionKey;
+import com.azure.core.cryptography.AsyncKeyEncryptionKeyResolver;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.common.implementation.BufferStagingArea;
+import com.azure.storage.common.implementation.UploadUtils;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import javax.crypto.Cipher;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.AES;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.AES_GCM_NO_PADDING;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.AES_KEY_SIZE_BITS;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.EMPTY_BUFFER;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.ENCRYPTION_PROTOCOL_V2;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.TAG_LENGTH;
+
+class DecryptorV2 extends Decryptor {
+    private static final ClientLogger LOGGER = new ClientLogger(DecryptorV2.class);
+
+    protected DecryptorV2(AsyncKeyEncryptionKeyResolver keyResolver, AsyncKeyEncryptionKey keyWrapper,
+        EncryptionData encryptionData) {
+        super(keyResolver, keyWrapper, encryptionData);
+    }
+
+    @Override
+    Flux<ByteBuffer> decrypt(Flux<ByteBuffer> encryptedFlux, EncryptedBlobRange encryptedBlobRange,
+        boolean padding, String requestUri, AtomicLong totalInputBytes, byte[] contentEncryptionKey) {
+        // Buffer an exact region with the nonce and tag
+        final int gcmEncryptionRegionLength = encryptionData.getEncryptedRegionInfo().getDataLength();
+        final int nonceLength = encryptionData.getEncryptedRegionInfo().getNonceLength();
+        BufferStagingArea stagingArea =
+            new BufferStagingArea(gcmEncryptionRegionLength + TAG_LENGTH + nonceLength,
+                gcmEncryptionRegionLength + TAG_LENGTH + nonceLength);
+
+        return UploadUtils.chunkSource(encryptedFlux,
+                new com.azure.storage.common.ParallelTransferOptions()
+                    .setBlockSizeLong((long) gcmEncryptionRegionLength
+                        + TAG_LENGTH + nonceLength))
+            .flatMapSequential(stagingArea::write)
+            .concatWith(Flux.defer(stagingArea::flush))
```

Comment:
did we check if chunkSource is needed here? It's a performance hit if not necessary.

Reply:
Working on that now.

Reply:
Removed the chunkSource and live tests look to be passing, so I think it's fine without.